def.bzl

```
@@ -176,21 +205,52 @@ def gazelle(name, **kwargs):
         kwargs.pop("args")
 
     visibility = kwargs.pop("visibility", default = None)
+    return name + "-runner", visibility
+
+def gazelle(name, **kwargs):
+    runner_name, visibility = _gazelle_kwargs_prepare(name, kwargs)
 
+    _gazelle_runner(
+        name = runner_name,
+        **kwargs
+    )
     tags_set = {t: "" for t in kwargs.pop("tags", [])}
     tags_set["manual"] = ""
     tags = [k for k in tags_set.keys()]
-    runner_name = name + "-runner"
-    _gazelle_runner(
-        name = runner_name,
+    native.sh_binary(
+        name = name,
         tags = tags,
+        srcs = [runner_name],
+        visibility = visibility,
+        deps = ["@bazel_tools//tools/bash/runfiles"],
+        data = kwargs["data"] if "data" in kwargs else [],
+    )
+
+def gazelle_test(name, no_sandbox = False, **kwargs):
+    if not no_sandbox:
+        fail("gazelle_test can only be used with no_sandbox = True.")
+    if not kwargs.get("workspace", ""):
+        fail("attr 'workspace' is required for gazelle_test")
+
+    runner_name, visibility = _gazelle_kwargs_prepare(name, kwargs)
+
+    _gazelle_test_runner(
+        name = runner_name,
         **kwargs
     )
-    native.sh_binary(
+
+    # Note: the "external" tag is a workaround for
+    # https://github.com/bazelbuild/bazel/issues/15516.
+    tags_set = {t: "" for t in kwargs.pop("tags", [])}
+    tags_set["no-sandbox"] = ""
+    tags_set["no-cache"] = ""
+    tags_set["no-external"] = ""
```

```suggestion
    tags_set["external"] = ""
```
based on the comment.